lvgl-codegen/src/lib.rs

```diff
@@ -0,0 +1,575 @@
+        Self { typ }
+    }
+
+    pub fn is_const(&self) -> bool {
+        self.typ.starts_with("const ")
+    }
+
+    pub fn is_str(&self) -> bool {
+        self.typ.ends_with("char *")
+    }
+}
+
+impl Rusty for LvType {
+    type Parent = LvArg;
+
+    fn code(&self, _parent: &Self::Parent) -> WrapperResult<TokenStream> {
+        match TYPE_MAPPINGS.get(self.typ.as_str()) {
+            Some(name) => {
+                let val = if self.is_str() {
+                    quote!(&str)
+                } else {
+                    let ident = format_ident!("{}", name);
+                    quote!(#ident)
+                };
+                Ok(quote! {
+                    #val
+                })
+            }
+            None => Err(WrapperError::Skip),
+        }
+    }
+}
+
+impl From<Type<'_>> for LvType {
+    fn from(ty: Type) -> Self {
+        Self::new(ty.get_display_name())
+    }
+}
+
+pub struct CodeGen {
+    functions: Vec<LvFunc>,
+    widgets: Vec<LvWidget>,
+}
+
+impl CodeGen {
+    pub fn new() -> CGResult<Self> {
+        let functions = Self::load_function_definitions()?;
+        let widgets = Self::extract_widgets(&functions)?;
+        Ok(Self { functions, widgets })
+    }
+
+    pub fn get_widgets(&self) -> &Vec<LvWidget> {
+        &self.widgets
+    }
+
+    fn extract_widgets(functions: &Vec<LvFunc>) -> CGResult<Vec<LvWidget>> {
+        let widget_names = Self::get_widget_names(functions);
+
+        let widgets = functions.iter().fold(HashMap::new(), |mut ws, f| {
+            for widget_name in &widget_names {
+                if f.name
+                    .starts_with(format!("{}{}", LIB_PREFIX, widget_name).as_str())
+                    && f.is_method()
+                {
+                    ws.entry(widget_name.clone())
+                        .or_insert_with(|| LvWidget {
+                            name: widget_name.clone(),
+                            methods: Vec::new(),
+                        })
+                        .methods
+                        .push(f.clone())
+                }
+            }
+            ws
+        });
+
+        Ok(widgets.values().map(|v| v.clone()).collect())
+    }
+
+    fn get_widget_names(functions: &Vec<LvFunc>) -> Vec<String> {
+        let reg = format!("^{}([^_]+)_create$", LIB_PREFIX);
+        let create_func = Regex::new(reg.as_str()).unwrap();
+
+        functions
+            .iter()
+            .filter(|e| create_func.is_match(e.name.as_str()) && e.args.len() == 2)
+            .filter_map(|f| {
+                Some(String::from(
+                    create_func
+                        .captures(f.name.as_str())
+                        .unwrap()
+                        .get(1)
+                        .unwrap()
+                        .as_str(),
+                ))
+            })
+            .collect::<Vec<_>>()
+    }
+
+    pub fn load_function_definitions() -> CGResult<Vec<LvFunc>> {
+        let clang = Clang::new()?;
```

Comment:
This line is where is clashes with `rust-bindgen`. It happens because `clang` has global state and so the Rust bindings to it prohibit to instantiate more than one per program. When `cargo` is building the `lvgl-rs` packages it uses the same process to build in different threads, so when building this package and bindgen we get errors.